<a href="https://colab.research.google.com/github/VeraMikhalchenko/Study/blob/main/HW1_Statistic_Base_Mikhalchenko_DSU_30.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Домашнее задание к лекции "Базовые понятия статистики"**

**Задание 1. Загрузка данных**\
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: \
surgery?,
Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [ ]:
import pandas as pd

In [ ]:
df_horses = pd.read_csv('horse_data.csv', header = None, na_values= '?')
df_horses.head()

0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0  2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1  1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2  2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3  1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4  2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   

    20   21   22  23     24  25  26  27  
0  NaN  NaN  2.0   2  11300   0   0   2  
1  2.0  2.0  3.0   2   2208   0   0   2  
2  NaN  NaN  1.0   2      0   0   0   1  
3  3.0  5.3  2.0   1   2208   0   0   1  
4  NaN  NaN  2.0   2   4300   0   0   2  

[5 rows x 28 columns]

In [ ]:
df_horses = df_horses.rename({0: 'surgery',1: 'age', 3: 'rectal_temperature', 
                              4: 'pulse', 5: 'respiratory_rate', 6: 'temperature_of_extremities', 
                              10: 'pain', 22:'outcome'}, 
                             axis=1)  
df_horses.head()

surgery  age        2  rectal_temperature  pulse  respiratory_rate  \
0      2.0    1   530101                38.5   66.0              28.0   
1      1.0    1   534817                39.2   88.0              20.0   
2      2.0    1   530334                38.3   40.0              24.0   
3      1.0    9  5290409                39.1  164.0              84.0   
4      2.0    1   530255                37.3  104.0              35.0   

   temperature_of_extremities    7    8    9  ...    18    19   20   21  \
0                         3.0  3.0  NaN  2.0  ...  45.0   8.4  NaN  NaN   
1                         NaN  NaN  4.0  1.0  ...  50.0  85.0  2.0  2.0   
2                         1.0  1.0  3.0  1.0  ...  33.0   6.7  NaN  NaN   
3                         4.0  1.0  6.0  2.0  ...  48.0   7.2  3.0  5.3   
4                         NaN  NaN  6.0  2.0  ...  74.0   7.4  NaN  NaN   

   outcome  23     24  25  26  27  
0      2.0   2  11300   0   0   2  
1      3.0   2   2208   0   0   2  
2      1.0   2      0   0   0   1  
3      2.0   1   2208   0   0   1  
4      2.0   2   4300   0   0   2  

[5 rows x 28 columns]

In [ ]:
df_horses = df_horses[['surgery', 'age', 'rectal_temperature', 'pulse', 'respiratory_rate', 'temperature_of_extremities', 'pain', 'outcome']]
df_horses.head()

surgery  age  rectal_temperature  pulse  respiratory_rate  \
0      2.0    1                38.5   66.0              28.0   
1      1.0    1                39.2   88.0              20.0   
2      2.0    1                38.3   40.0              24.0   
3      1.0    9                39.1  164.0              84.0   
4      2.0    1                37.3  104.0              35.0   

   temperature_of_extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

**Задание 2. Первичное изучение данных**

Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

In [ ]:
df_horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     299 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            242 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [ ]:
df_horses.describe()

surgery         age  rectal_temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

Напишем функцию для расчета базовых статистик:

In [ ]:
def statistics_base(data):
    """ Функция находит с помощью pandas базовые статистики:
    - среднее арифметическое
    - медиану
    - моду
    - стандартное отклонение
    - дисперсию
    """
    data_mean = data.mean()
    data_median = data.median()
    data_mode = data.mode()[0]
    data_std = data.std()
    data_var = data.var()

    print(f'среднее: {data_mean}')
    print(f'медиана: {data_median}')
    print(f'мода: {data_mode}')
    print(f'стандартное отклонение: {data_std}')
    print(f'дисперсия: {data_var}')
    
    return

Напишем функцию для поиска выбросов через межквартильный размах:

In [ ]:
def find_anomalies_iqr(data):
    """ Функция находит выбросы через межквартильный размах.
    """
    outliers = []
    q1 = data.quantile(0.25)
    q3 = data.quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)

    for outlier in data:
        if outlier > upper_bound or outlier < lower_bound:
            outliers.append(outlier)
    
    print('Выбросы, найденные с помощью межквартильного размаха:')
    return outliers

Напишем функцию для поиска выбросов через стандартное отклонение:

In [ ]:
def find_anomalies_std(data):
    """Функция находит выбросы с помощью стандартного отклонения. 
     Если у вас есть какая-либо точка данных, которая более чем 
     в 3 раза превышает стандартное отклонение, то эти точки, 
     скорее всего, будут аномальными или выбросами.
    """
    anomalies = []
    data_std = data.std()
    data_mean = data.mean()
    anomaly_cut_off = data_std * 3
    
    lower_bound  = data_mean - anomaly_cut_off 
    upper_bound = data_mean + anomaly_cut_off
   
    for outlier in data:
        if outlier > upper_bound or outlier < lower_bound:
            anomalies.append(outlier)
    print('Выбросы, найденные с помощью стандартного отклонения:')
    return anomalies

**surgery:**\
          1 = Yes, it had surgery;\
          2 = It was treated without surgery \

**surgery** - качественный бинарный атрибут, свидетельствующий о том делали оперцию лошади или нет. Для бинарного атрибута нет смысла рассчитывать базовые метрики. Посмотрим, как распределены значения в данном атрибуте. (1 NaN)

In [ ]:
df_horses['surgery'].value_counts()

1.0    180
2.0    119
Name: surgery, dtype: int64

**age:** \
          1 = Adult horse;\
          2 = Young (< 6 months)\
**age** - качественный бинариный атрибут. (0 NaN)

In [ ]:
df_horses['age'].value_counts()

1    276
9     24
Name: age, dtype: int64

В описании данных значение столбца **age** может быть равно либо 1 (взрослая лошадь), либо 2 (молодая лошадь, возраст < 6 месяцев). В данной таблице есть значения 1 или 9. Что именно означает 9 — не понятно. По другим данным так же сложно это выяснить. Можно предположить, что это месяцы, тогда все лошади в данном датасете взрослые, можно предположить, что это ошибочное написание '2', тогда 9 - это молодые лошади. Дополнительных атрибутов, по которым мы смогли бы судить о возрасте, я не нашла. Разве что пульс, который обычно у более молодых особей более частый, чем у взрослых (https://megaobuchalka.ru/2/18407.html). Но довольно часто на пульс влияет и состояние здоровья, и температура тела.

Найдём среднее значение показателя пульса у лощадей с возрастом '9' и возрастом '1' и сравним.

In [ ]:
df_horses.pulse[df_horses['age'] == 9].mean()

123.36363636363636

In [ ]:
df_horses.pulse[df_horses['age'] == 1].mean()

67.45669291338582

Если экспертно оценивать эти значения, то разница существенная для пульса, но статистически значимость этих различий я не могу определить (возможно, тут может помочь U-тест Манна - Уитни). И всё же предположим, что 9 - это молодые лошади.\
Для того, чтобы оценить необходимость замены "непонятных" данных в столбце с возрастом, хочется иметь представление о задаче, для которой могут понадобиться эти данные. Например, нам надо спрогнозировать, какая учесть может ждать лошадь: жизнь или смерть (с вариантом эвтаназии).\
Для такой задачи показатель возраста может быть очень существенным, и в реальной жизни необходимо было бы уточнить, что это за таинственная 9-ка у тех, кто собирал данные. 

Поскольку мы предположили, что возраст связан с пульсом, и 9 - это молодые лошади, то заменим все 9-ки на 2.

In [ ]:
df_horses.loc[(df_horses.age == 9), 'age'] = 2

df_horses.age[df_horses['age'] == 2].count() #проверяю количество изменённых значений

24

**rectal_temperature:**\
          - linear\
          - in degrees celsius\
          - An elevated temp may occur due to infection\
          - temperature may be reduced when the animal is in late shock\
          - normal temp is 37.8\
          - this parameter will usually change as the problem progresses\
               eg. may start out normal, then become elevated because of
                   the lesion, passing back through the normal range as the
                   horse goes into shock\
**rectal_temperature** - количественный атрибут (60 NaN).

Проверим, нет ли каких-нибудь ошибочных данных.

In [ ]:
df_horses['rectal_temperature'].value_counts()

38.0    26
38.5    19
38.3    18
37.8    17
38.2    16
38.1    12
37.5    12
38.6    12
38.4    11
37.9     7
37.2     7
37.6     7
37.7     7
38.7     7
37.3     6
38.8     6
39.2     5
39.5     4
39.3     4
38.9     4
39.0     4
39.4     3
37.1     3
37.4     3
40.3     2
37.0     2
39.1     2
36.5     2
36.9     1
40.8     1
36.6     1
36.1     1
35.4     1
36.0     1
36.8     1
39.6     1
39.9     1
36.4     1
39.7     1
40.0     1
Name: rectal_temperature, dtype: int64

Все встречающиеся значения выглядят соотвествующими показателю температуры.

Рассчитаем базовые метрики:

In [ ]:
statistics_base(df_horses['rectal_temperature'])

среднее: 38.16791666666667
медиана: 38.2
мода: 38.0
стандартное отклонение: 0.7322886641121578
дисперсия: 0.5362466875871686


Предположим, что распределение данных по этому атрибуту нормальное и определим выбросы через межквартильный размах.

In [ ]:
find_anomalies_iqr(df_horses['rectal_temperature'])

Выбросы, найденные с помощью межквартильного размаха:


[39.9,
 35.4,
 40.3,
 39.7,
 36.4,
 40.3,
 39.6,
 36.5,
 36.0,
 36.1,
 36.6,
 40.8,
 40.0,
 36.5]

Определим выбросы через стандартное отклонение:

In [ ]:
find_anomalies_std(df_horses['rectal_temperature'])

Выбросы, найденные с помощью стандартного отклонения:


[35.4, 40.8]

Если оценивать эти показатели выбросов с помощью зравого смысла (что, скорее всего, ошибочно в статистике), то их следует принимать во внимание в расчетах. Это естественные выбросы.

**pulse:**\
          - linear\
          - the heart rate in beats per minute\
          - is a reflection of the heart condition: 30 -40 is normal for adults\
          - rare to have a lower than normal rate although athletic horses
            may have a rate of 20-25\
          - animals with painful lesions or suffering from circulatory shock
            may have an elevated heart rate\
**pulse** - количественный атрибут(24 NaN).

Посмотрим, какие значения встречаются по данному атрибуту.

In [ ]:
df_horses['pulse'].value_counts()

48.0     28
60.0     25
40.0     18
52.0     12
88.0     12
44.0     11
100.0    11
72.0     11
120.0    10
42.0     10
64.0      8
96.0      8
84.0      8
80.0      8
54.0      7
56.0      7
50.0      6
66.0      6
104.0     5
68.0      5
92.0      4
70.0      4
76.0      4
108.0     3
86.0      3
78.0      3
90.0      3
36.0      3
112.0     3
114.0     2
45.0      2
140.0     2
124.0     2
150.0     2
130.0     2
30.0      2
132.0     1
129.0     1
184.0     1
75.0      1
65.0      1
164.0     1
136.0     1
49.0      1
98.0      1
38.0      1
160.0     1
46.0      1
128.0     1
146.0     1
110.0     1
82.0      1
Name: pulse, dtype: int64

На первый взгляд, все данные соответствуют показателю пульса.

Рассчитаем базовые метрики:

In [ ]:
statistics_base(df_horses['pulse'])

среднее: 71.91304347826087
медиана: 64.0
мода: 48.0
стандартное отклонение: 28.630556660735
дисперсия: 819.7087747035573


Выбросы:

In [ ]:
find_anomalies_iqr(df_horses['pulse'])

Выбросы, найденные с помощью межквартильного размаха:


[164.0, 150.0, 160.0, 184.0, 150.0]

In [ ]:
find_anomalies_std(df_horses['pulse'])

Выбросы, найденные с помощью стандартного отклонения:


[164.0, 160.0, 184.0]

Согласно здравому смыслу, эти показатели выбросов могут быть скорее свидетельством особого состояния здоровья лошади, чем ошибки, поэтому следует их оставить как есть. Это естественные выбросы

**respiratory_rate:**\
          - linear\
          - normal rate is 8 to 10\
          - usefulness is doubtful due to the great fluctuations\
**respiratory_rate** - количественный атрибут (58 NaN).

Посмотрим, какие данные встречаются:

In [ ]:
df_horses['respiratory_rate'].value_counts()

20.0    28
24.0    27
16.0    22
12.0    19
30.0    19
40.0    17
36.0    16
28.0    13
32.0    11
18.0     8
48.0     6
60.0     4
14.0     4
42.0     3
68.0     3
10.0     3
44.0     3
35.0     3
80.0     3
9.0      2
51.0     2
90.0     2
70.0     2
22.0     2
96.0     2
21.0     2
84.0     2
50.0     2
15.0     1
72.0     1
66.0     1
34.0     1
52.0     1
8.0      1
23.0     1
58.0     1
13.0     1
26.0     1
88.0     1
25.0     1
Name: respiratory_rate, dtype: int64

Рассчитаем базовые метрики:

In [ ]:
statistics_base(df_horses['respiratory_rate'])

среднее: 30.417355371900825
медиана: 24.5
мода: 20.0
стандартное отклонение: 17.64223138513467
дисперсия: 311.2483282466308


Выбросы:

In [ ]:
find_anomalies_iqr(df_horses['respiratory_rate'])

Выбросы, найденные с помощью межквартильного размаха:


[84.0,
 96.0,
 72.0,
 80.0,
 80.0,
 68.0,
 96.0,
 66.0,
 68.0,
 90.0,
 80.0,
 70.0,
 88.0,
 84.0,
 68.0,
 90.0,
 70.0]

In [ ]:
find_anomalies_std(df_horses['respiratory_rate'])

Выбросы, найденные с помощью стандартного отклонения:


[84.0, 96.0, 96.0, 90.0, 88.0, 84.0, 90.0]

Это естественные выбросы. Учитывая, что в описании датасета указано, что норма находится в диапаоне 8-10 по этому атрибуту, не очень понятно, что делать с этими выбросами в дальнейшем и со всем атрибутом в целом. Возможно, от этого атрибута придётся отказаться в зависимости от задачи.

**temperature_of_extremities:**\
          - a subjective indication of peripheral circulation\
          - possible values:\
               1 = Normal\
               2 = Warm\
               3 = Cool\
               4 = Cold\
          - cool to cold extremities indicate possible shock\
          - hot extremities should correlate with an elevated rectal temp\
**temperature_of_extremities** - качественный порядковый атрибут (56 NaN).

In [ ]:
df_horses['temperature_of_extremities'].value_counts()

3.0    109
1.0     78
2.0     30
4.0     27
Name: temperature_of_extremities, dtype: int64

**pain:**\
          - possible values:\
               1 = alert, no pain\
               2 = depressed\
               3 = intermittent mild pain\
               4 = intermittent severe pain\
               5 = continuous severe pain\
          - should NOT be treated as a ordered or discrete variable!\
          - In general, the more painful, the more likely it is to require
            surgery\
          - prior treatment of pain may mask the pain level to some extent\
**pain** - качественный номинальный атрибут (55 NaN).

In [ ]:
df_horses['pain'].value_counts()

3.0    67
2.0    59
5.0    42
4.0    39
1.0    38
Name: pain, dtype: int64


**outcome:**\
          - what eventually happened to the horse?\
          - possible values:\
               1 = lived\
               2 = died\
               3 = was euthanized\
**outcome** - качественный атрибут (1 NaN)

In [ ]:
df_horses['outcome'].value_counts()

1.0    178
2.0     77
3.0     44
Name: outcome, dtype: int64

**Задание 3. Работа с пропусками**

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

В случае с медицинскими показателями наиболее правильным представляется решение просто удалить все строки с пропусками, потому что не очевидны механизмы формирования пропусков, и природа данных слишком индивидуальна. В учебных целях редположим, что в нашем случае это MCAR.

Проанализируем пропуски по каждому атрибуту:

In [ ]:
df_horses.isna().sum()

surgery                        1
age                            0
rectal_temperature            60
pulse                         24
respiratory_rate              58
temperature_of_extremities    56
pain                          55
outcome                        1
dtype: int64

В описании датасета указано, что столбец respiratory_rate не информативен в силу большого разброса и там много пропусков. Можно удалить весь этот столбец.

In [ ]:
df_horses = df_horses.drop(['respiratory_rate'], axis=1)
df_horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     299 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   temperature_of_extremities  244 non-null    float64
 5   pain                        245 non-null    float64
 6   outcome                     299 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 16.5 KB


В строках, где 3 из 7 атрибутов с пропусками - малоинформативны. Оставим строки, в которых как минимум 5/7 атрибутов заполнены

In [ ]:
df_horses_dropna = df_horses.dropna(thresh=5)
df_horses_dropna

surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
0        2.0    1                38.5   66.0                         3.0   
1        1.0    1                39.2   88.0                         NaN   
2        2.0    1                38.3   40.0                         1.0   
3        1.0    2                39.1  164.0                         4.0   
4        2.0    1                37.3  104.0                         NaN   
..       ...  ...                 ...    ...                         ...   
295      1.0    1                 NaN  120.0                         4.0   
296      2.0    1                37.2   72.0                         3.0   
297      1.0    1                37.5   72.0                         4.0   
298      1.0    1                36.5  100.0                         3.0   
299      1.0    1                37.2   40.0                         NaN   

     pain  outcome  
0     5.0      2.0  
1     3.0      3.0  
2     3.0      1.0  
3     2.0      2.0  
4     NaN      2.0  
..    ...      ...  
295   2.0      3.0  
296   4.0      3.0  
297   4.0      2.0  
298   3.0      1.0  
299   NaN      3.0  

[288 rows x 7 columns]

Посмотрим, как изменились показатели пропусков

In [ ]:
df_horses_dropna.isna().sum()

surgery                        1
age                            0
rectal_temperature            48
pulse                         15
temperature_of_extremities    44
pain                          44
outcome                        1
dtype: int64

Поскольку атрибут age не содержит пропусков, мы можем на основе этого атрибута заполнить пропуски в атрибуте pulse (основываясь на том, что тем старше лошадь, тем ниже пульс). Выбираем медиану, потому что есть выбросы.

In [ ]:
print(df_horses_dropna.groupby('age')['pulse'].median())
fill_na = df_horses_dropna
fill_na['pulse'] = fill_na['pulse'].fillna(fill_na.groupby('age')['pulse'].transform('median'))

age
1     60.0
2    124.0
Name: pulse, dtype: float64


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
fill_na.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 288 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     287 non-null    float64
 1   age                         288 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       288 non-null    float64
 4   temperature_of_extremities  244 non-null    float64
 5   pain                        244 non-null    float64
 6   outcome                     287 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.0 KB


In [ ]:
fill_na

surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
0        2.0    1                38.5   66.0                         3.0   
1        1.0    1                39.2   88.0                         NaN   
2        2.0    1                38.3   40.0                         1.0   
3        1.0    2                39.1  164.0                         4.0   
4        2.0    1                37.3  104.0                         NaN   
..       ...  ...                 ...    ...                         ...   
295      1.0    1                 NaN  120.0                         4.0   
296      2.0    1                37.2   72.0                         3.0   
297      1.0    1                37.5   72.0                         4.0   
298      1.0    1                36.5  100.0                         3.0   
299      1.0    1                37.2   40.0                         NaN   

     pain  outcome  
0     5.0      2.0  
1     3.0      3.0  
2     3.0      1.0  
3     2.0      2.0  
4     NaN      2.0  
..    ...      ...  
295   2.0      3.0  
296   4.0      3.0  
297   4.0      2.0  
298   3.0      1.0  
299   NaN      3.0  

[288 rows x 7 columns]

Согласно описанию датасета, показатели temperature_of_extremities выше, если повышены показатели rectal_temperature. Вопрос в том, что является повышенными показателями rectal_temperature. Предположим, что это те данные, которые находятся выше 3 квартили.

In [ ]:
q3 = fill_na['rectal_temperature'].quantile(0.75)
q3

38.5

Везде, где temperature_of_extremities высокая (2), проставим высокие значения rectal_temperature.

Найдем медиану по высоким показателям rectal_temperature:

In [ ]:
fill_na['rectal_temperature'][fill_na['rectal_temperature'] > 38.5].median()

38.95

Проставим это значение везде, где temperature_of_extremities = 2

In [ ]:
fill_na['rectal_temperature'][fill_na['temperature_of_extremities'] == 2] = fill_na['rectal_temperature'].fillna(38.5)
fill_na.isna().sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:8870: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


surgery                        1
age                            0
rectal_temperature            44
pulse                          0
temperature_of_extremities    44
pain                          44
outcome                        1
dtype: int64

И наоборот, везде, где rectal_temperature высокая, проставим temperature_of_extremities = 2:

In [ ]:
fill_na['temperature_of_extremities'][fill_na['rectal_temperature'] > 38.5] = fill_na['temperature_of_extremities'].fillna(2)
fill_na.isna().sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


surgery                        1
age                            0
rectal_temperature            44
pulse                          0
temperature_of_extremities    35
pain                          44
outcome                        1
dtype: int64

Атрибут pain является качественным номинальным показателем боли (довольно интересно, как её можно оценить у животного). Можно поробовать выделить отдельную категорию с пропусками: 0 - боль не определена.

Заменим все пропуски в атрибуте pain на 0:

In [ ]:
fill_na['pain'] = fill_na['pain'].fillna(0)
fill_na.isna().sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


surgery                        1
age                            0
rectal_temperature            44
pulse                          0
temperature_of_extremities    35
pain                           0
outcome                        1
dtype: int64

Остальные строки с пропущенными значениями просто удалим, поскольку мне сложно придумать еще основания для заполнения их какими-либо значениями. Возможно, можо было бы посмотреть на статистически значимые взяимосвязи между атрибутами, и на основе этого заполнить недостающие значения.

In [ ]:
final_df_horses = fill_na.dropna()
final_df_horses.isna().sum()

surgery                       0
age                           0
rectal_temperature            0
pulse                         0
temperature_of_extremities    0
pain                          0
outcome                       0
dtype: int64

In [ ]:
final_df_horses

surgery  age  rectal_temperature  pulse  temperature_of_extremities  \
0        2.0    1                38.5   66.0                         3.0   
1        1.0    1                39.2   88.0                         2.0   
2        2.0    1                38.3   40.0                         1.0   
3        1.0    2                39.1  164.0                         4.0   
5        2.0    1                38.5   60.0                         2.0   
..       ...  ...                 ...    ...                         ...   
292      1.0    1                37.0   66.0                         1.0   
294      2.0    1                38.5   40.0                         1.0   
296      2.0    1                37.2   72.0                         3.0   
297      1.0    1                37.5   72.0                         4.0   
298      1.0    1                36.5  100.0                         3.0   

     pain  outcome  
0     5.0      2.0  
1     3.0      3.0  
2     3.0      1.0  
3     2.0      2.0  
5     2.0      1.0  
..    ...      ...  
292   4.0      2.0  
294   2.0      1.0  
296   4.0      3.0  
297   4.0      2.0  
298   3.0      1.0  

[210 rows x 7 columns]

Мы потеряли почти 40% данных.

Сравним, как поменялись базовые метрики:

In [ ]:
print('rectal_temperature до замены и удаления пропусков:')
statistics_base(df_horses['rectal_temperature'])

rectal_temperature до замены и удаления пропусков:
среднее: 38.16791666666667
медиана: 38.2
мода: 38.0
стандартное отклонение: 0.7322886641121578
дисперсия: 0.5362466875871686


In [ ]:
print('rectal_temperature после замены и удаления пропусков:')
statistics_base(final_df_horses['rectal_temperature'])

rectal_temperature после замены и удаления пропусков:
среднее: 38.23476190476191
медиана: 38.2
мода: 38.0
стандартное отклонение: 0.7345741796007175
дисперсия: 0.5395992253360672


Почти ничего не поменялось.

In [ ]:
print('pulse до замены и удаления пропусков:')
statistics_base(df_horses['pulse'])

pulse до замены и удаления пропусков:
среднее: 71.91304347826087
медиана: 64.0
мода: 48.0
стандартное отклонение: 28.630556660735
дисперсия: 819.7087747035573


In [ ]:
print('pulse после замены и удаления пропусков:')
statistics_base(final_df_horses['pulse'])

pulse после замены и удаления пропусков:
среднее: 70.94761904761904
медиана: 60.0
мода: 60.0
стандартное отклонение: 28.582661439652504
дисперсия: 816.9685349737981


Сильно изменились показатели моды, медианы и дисперсии. Если мы будем строить модель логистической регрессии, то изменение показателя дисперсии будет искажать картину.

In [ ]:
df_horses['surgery'].value_counts()

1.0    171
2.0    116
Name: surgery, dtype: int64

In [ ]:
final_df_horses['surgery'].value_counts()

1.0    102
2.0     83
Name: surgery, dtype: int64

Пропорция приблизительно та же.

In [ ]:
df_horses['age'].value_counts()

1    266
2     22
Name: age, dtype: int64

In [ ]:
final_df_horses['age'].value_counts()

1    170
2     15
Name: age, dtype: int64

Пропоция приблизительно та же.

In [ ]:
df_horses['temperature_of_extremities'].value_counts()

3.0    109
1.0     78
2.0     30
4.0     27
Name: temperature_of_extremities, dtype: int64

In [ ]:
final_df_horses['temperature_of_extremities'].value_counts()

3.0    73
1.0    68
2.0    28
4.0    16
Name: temperature_of_extremities, dtype: int64

Порядок значений остался тот же. Пропорция немного изменилась.

In [ ]:
df_horses['pain'].value_counts()

3.0    67
2.0    59
5.0    42
4.0    38
1.0    38
Name: pain, dtype: int64

In [ ]:
final_df_horses['pain'].value_counts()

3.0    57
2.0    45
1.0    35
4.0    26
5.0    22
Name: pain, dtype: int64

Распределение показателей изменилось. Так, например, отсутствие боли (1) переместилось на третье место. Любопытно, что категория, которая была создана с целью заполнить пропуски '0 - боль не определена', исчезла после того, как мы удалили все пропуски.

In [ ]:
df_horses['outcome'].value_counts()

1.0    171
2.0     74
3.0     42
Name: outcome, dtype: int64

In [ ]:
final_df_horses['outcome'].value_counts()

1.0    117
2.0     44
3.0     24
Name: outcome, dtype: int64

Распределение данных атрибута не сильно поменялось.